### ISSUE TITLE
[libcxx] Miscompilation of string code when using new-struct-path-tbaa

### ISSUE THREAD

**Issue by brunodefraine** [llvm/llvm-project#55384]
When using libcxx and clang 14, we observe a miscompilation of the following simple `std::string` code:

```cpp
std::string r;
...
r = std::string("Hello, ") + "world";
```

We expect string r to be equal to "Hello, world" but we end up with "H" followed by 11 garbage characters, depending on the earlier contents of r. This is a [godbolt demo](https://godbolt.org/z/qdo7WPEzz). The only non-trivial factor to trigger the wrong behavior, is the use of the new TBAA struct path format (using clang option `-new-struct-path-tbaa`).

# Analysis

After appending to the left-hand side string, the statement triggers a move constructor and move assignment to get the result in place. Each move consists of a copy of the string representation, followed by a zero operation that clears the representation of the stolen string. It seems these memory access operations do not have proper TBAA information, leading to incorrect memory optimization. 

## TBAA tags

In the [LLVM IR for the move assignment operator and move constructor](https://godbolt.org/z/fdver7dd5), we can find back the TBAA tags for the memory operations. These tags should be compatible as they may refer to overlapping memory access. The internal representation of a basic_string is the following structure:

```cpp
    union __ulx{__long __lx; __short __lxx;};

    enum {__n_words = sizeof(__ulx) / sizeof(size_type)};

    struct __raw
    {
        size_type __words[__n_words];
    };

    struct __rep
    {
        union
        {
            __long  __l;
            __short __s;
            __raw   __r;
        };
    };

    __compressed_pair<__rep, allocator_type> __r_;
```

Type-punning using a union violates strict aliasing, but is supported by e.g. GCC [provided the memory is accessed through the union type](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#Type-punning). Also in clang, access using a union is normally given the tag of "omnipotent char".

The move assignment operator is doing (in `__move_assign`):

```cpp
  __r_.first() = __str.__r_.first();
  __str.__set_short_size(0);
```

The assignment at the level of `__rep` gives a `memcpy` intrinsic with TBAA base and access tag of `__rep`, which has a member (the anonymous union) of omnipotent char. `__set_short_size` uses the access expression `__r_.first().__s.__size_`, the union access has base and access tag of omnipotent char.

The move constructor is doing:

```cpp
    : __r_(_VSTD::move(__str.__r_))
{
    __str.__zero();
```

The move at the level of the `__compressed_pair` gives a `memcpy` intrinsic with TBAA base and access tag of the `__compressed_pair`, but this structure has no members in the TBAA info, **this seems incomplete**:

```llvm
!8 = !{!9, !9, i64 0, i64 24}
!9 = !{!6, i64 24, !"_ZTSNSt3__117__compressed_pairINS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE5__repES5_EE"}
```

The function `__zero` is doing:

```cpp
    void __zero() _NOEXCEPT
        {
            size_type (&__a)[__n_words] = __r_.first().__r.__words;
            for (unsigned __i = 0; __i < __n_words; ++__i)
                __a[__i] = 0;
        }
```

The store of the zero value in the loop body is given TBAA base and access tag of `long`, **this is clearly wrong**. However, function `__zero` violates the constraint "memory is accessed through the union type" from the GCC manual (cited above). The use of the intermediate reference variable `a` is equivalent to the counter-example (using an intermediate pointer variable `int* ip`) given there. If the source code is changed to remove the use of the reference variable, the base and access tag is omnipotent char, as expected.

# Workaround

Changing the code of `__move_assign` do the assignment at the level of `__raw`:

```cpp
  __r_.first().__r = __str.__r_.first().__r;
```

Gives a memcpy with TBAA tag of omnipotent char (since we are accessing a union member) and prevents the incorrect memory optimization. Presumably because this forces to alias with the `__compressed_pair` (but as noted above, that tag seems incomplete).

# Credits

Bhramar Bhushan Vatsa, Wouter Vermaelen and Jeroen Dobbelaere contributed to this bug report.

**Comment by philnik777** [comment]
This is very fascinating and I understand about 10% of it, but this sounds to me like a clang or LLVM bug, not a libc++ one. Or did I miss something? If I'm not mistaken, could you add the correct labels (assuming you know where the bug is)? If you don't know you could try it with `clang` and see if someone knows better over there.

**Comment by fhahn** [comment]
I added a new TBAA label, as we already have a coupe of issues related to TBAA. cc @efriedma-quic @rjmccall 

> Type-punning using a union violates strict aliasing, but is supported by e.g. GCC [provided the memory is accessed through the union type](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#Type-punning). Also in clang, access using a union is normally given the tag of "omnipotent char".

I am not aware of Clang providing any documented guarantees if strict aliasing is violated, but this area is not documented perfectly at the moment.

**Comment by philnik777** [comment]
I removed libc++, since this has now found a new home.

**Comment by efriedma-quic** [comment]
From the description, this sounds like a bug in the libc++ source code?  Or if it isn't, we need a reduced testcase that doesn't use libc++.

**Comment by philnik777** [comment]
If it's a libc++ then it should be resolved in trunk. We've changed the string internals a lot in the last few months and the loop in `__zero` doesn't exist anymore. The data structure also changed.

**Comment by fhahn** [comment]
> If it's a libc++ then it should be resolved in trunk. We've changed the string internals a lot in the last few months and the loop in __zero doesn't exist anymore. The data structure also changed.

I am confused, are you saying this is fixed in current `main` and only a problem in `14.0`? If that's the case, I think we can close the issue unless there is a patch that could be back-ported.

**Comment by philnik777** [comment]
@fhahn I'm saying that if this were a libc++ bug as described it would be fixed in trunk. Looking at the godbolt link it still mis-compiles, so it's probably not a libc++ bug.

**Comment by brunodefraine** [comment]
@philnik777 Thanks for pointing out that `__zero` has been recently rewritten. I still observed that the TBAA tag of the `__compressed_pair` seems incomplete (I presume this class uses a special structure to exploit the empty base optimization). I'll try to make a reduced test case independent from libcxx.

**Comment by brunodefraine** [comment]
@efriedma-quic This is a self-contained, bare-bones (but still recognizable) version of the test case in ~100 lines:
https://godbolt.org/z/Gq11WKnrv

The program should return 0 (not 1). The new implementation of `__zero` does not resolve the issue.

**Comment by philnik777** [comment]
Here is a version that is a bit shorter: https://godbolt.org/z/43qGEco97

**Comment by efriedma-quic** [comment]
I agree this is a clang issue.  Reduced a bit more:

```
#include <cassert>
struct __short {
    char __data_[23];
};
struct __compressed_pair : public __short {};
struct string {
    __compressed_pair __r_;
    string(const char* __s) {
        volatile unsigned __sz = 3;
        char* __p = &__r_.__data_[0];
        for (unsigned i = 0; i < __sz; ++i)
            __p[i] = __s[i];
        __p[__sz] = char();
    }
    string& operator=(const string& __str) {
        static_cast<__short&>(__r_) = static_cast<const __short&>(__str.__r_);
        return *this;
    }
};
string z(string&& x) { return x; }
int main() {
    string r("def");
    r = z(string("abc"));
    assert(r.__r_.__data_[1] == 'b');
}
```

**Comment by brunodefraine** [comment]
> I agree this is a clang issue. Reduced a bit more:

Sweet! This is the generated code (obtained with `-Xclang -disable-llvm-passes -emit-llvm -g0`): https://godbolt.org/z/K4K94oc6P

I still think the bug is in the TBAA info for `__compressed_pair`: it does not record what it contains from its parent class:

```
!12 = !{!8, i64 23, !"_ZTS17__compressed_pair"}
```

The example compiles fine if I modify this to include a `__short` (tag `!19`):

```
!12 = !{!8, i64 23, !"_ZTS17__compressed_pair", !19, i64 0, i64 23}
```

**Comment by brunodefraine** [comment]
Submitted a fix as https://reviews.llvm.org/D126956

**Comment by tambry** [comment]
Fixed by cdc59e2202c11a6a5dfd2ec83531523c58eaae45.

**Comment by brunodefraine** [comment]
Not that the commit has been reverted pending a fix for the problems that it caused in the multistage builds.
